File: internal/plugins/jest/rules/prefer_to_have_been_called/prefer_to_have_been_called_test.go
package prefer_to_have_been_called_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/prefer_to_have_been_called"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestPreferToHaveBeenCalledRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&prefer_to_have_been_called.PreferToHaveBeenCalledRule,
		[]rule_tester.ValidTestCase{
			{Code: `expect(method.mock.calls).toHaveLength;`},
			{Code: `expect(method.mock.calls).toHaveLength(0);`},
			{Code: `expect(method).toHaveBeenCalledTimes(1)`},
			{Code: `expect(method).not.toHaveBeenCalledTimes(x)`},
			{Code: `expect(method).not.toHaveBeenCalledTimes(1)`},
			{Code: `expect(method).not.toHaveBeenCalledTimes(...x)`},
			{Code: `expect(a);`},
			{Code: `expect(method).not.resolves.toHaveBeenCalledTimes(0);`},
			{Code: `expect(method).toBe([])`},
			{Code: `expect(fn.mock.calls).toEqual([])`},
			{Code: `expect(fn.mock.calls).toContain(1, 2, 3)`},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code:   `expect(method).toBeCalledTimes(0);`,
				Output: []string{`expect(method).not.toHaveBeenCalled();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 16},
				},
			},
			{
				Code:   `expect(method).not.toBeCalledTimes(0);`,
				Output: []string{`expect(method).toHaveBeenCalled();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 20},
				},
			},
			{
				Code:   `expect(method).toHaveBeenCalledTimes(0);`,
				Output: []string{`expect(method).not.toHaveBeenCalled();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 16},
				},
			},
			{
				Code:   `expect(method).not.toHaveBeenCalledTimes(0);`,
				Output: []string{`expect(method).toHaveBeenCalled();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 20},
				},
			},
			{
				Code:   `expect(method).not.toHaveBeenCalledTimes(0, 1, 2);`,
				Output: []string{`expect(method).toHaveBeenCalled();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 20},
				},
			},
			{
				Code:   `expect(method).resolves.toHaveBeenCalledTimes(0);`,
				Output: []string{`expect(method).resolves.not.toHaveBeenCalled();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 25},
				},
			},
			{
				Code:   `expect(method).rejects.not.toHaveBeenCalledTimes(0);`,
				Output: []string{`expect(method).rejects.toHaveBeenCalled();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 28},
				},
			},
			{
				Code:   `expect(method).toBeCalledTimes(0 as number);`,
				Output: []string{`expect(method).not.toHaveBeenCalled();`},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "preferMatcher", Line: 1, Column: 16},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/prefer-to-have-been-called.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('prefer-to-have-been-called', {} as never, {
  valid: [
    { code: 'expect(method.mock.calls).toHaveLength;' },
    { code: 'expect(method.mock.calls).toHaveLength(0);' },
    { code: 'expect(method).toHaveBeenCalledTimes(1)' },
    { code: 'expect(method).not.toHaveBeenCalledTimes(x)' },
    { code: 'expect(method).not.toHaveBeenCalledTimes(1)' },
    { code: 'expect(method).not.toHaveBeenCalledTimes(...x)' },
    { code: 'expect(a);' },
    { code: 'expect(method).not.resolves.toHaveBeenCalledTimes(0);' },
    { code: 'expect(method).toBeCalledTimes(0!);' },
    { code: 'expect(method).toBeCalledTimes(0 satisfies number);' },
    { code: 'expect(method).toBe([])' },
    { code: 'expect(fn.mock.calls).toEqual([])' },
    { code: 'expect(fn.mock.calls).toContain(1, 2, 3)' },
  ],
  invalid: [
    {
      code: 'expect(method).toBeCalledTimes(0);',
      output: 'expect(method).not.toHaveBeenCalled();',
      errors: [{ messageId: 'preferMatcher', column: 16, line: 1 }],
    },
    {
      code: 'expect(method).not.toBeCalledTimes(0);',
      output: 'expect(method).toHaveBeenCalled();',
      errors: [{ messageId: 'preferMatcher', column: 20, line: 1 }],
    },
    {
      code: 'expect(method).toHaveBeenCalledTimes(0);',
      output: 'expect(method).not.toHaveBeenCalled();',
      errors: [{ messageId: 'preferMatcher', column: 16, line: 1 }],
    },
    {
      code: 'expect(method).not.toHaveBeenCalledTimes(0);',
      output: 'expect(method).toHaveBeenCalled();',
      errors: [{ messageId: 'preferMatcher', column: 20, line: 1 }],
    },
    {
      code: 'expect(method).not.toHaveBeenCalledTimes(0, 1, 2);',
      output: 'expect(method).toHaveBeenCalled();',
      errors: [{ messageId: 'preferMatcher', column: 20, line: 1 }],
    },

    {
      code: 'expect(method).resolves.toHaveBeenCalledTimes(0);',
      output: 'expect(method).resolves.not.toHaveBeenCalled();',
      errors: [{ messageId: 'preferMatcher', column: 25, line: 1 }],
    },
    {
      code: 'expect(method).rejects.not.toHaveBeenCalledTimes(0);',
      output: 'expect(method).rejects.toHaveBeenCalled();',
      errors: [{ messageId: 'preferMatcher', column: 28, line: 1 }],
    },

    {
      code: 'expect(method).toBeCalledTimes(0 as number);',
      output: 'expect(method).not.toHaveBeenCalled();',
      errors: [{ messageId: 'preferMatcher', column: 16, line: 1 }],
    },
  ],
});
